xrtl/gfx/es3/es3_swap_chain.cc

```diff
@@ -14,11 +14,201 @@
+
+  // Allocate framebuffers we'll use for copying.
+  framebuffers_.resize(image_count);
+  glGenFramebuffers(static_cast<GLsizei>(framebuffers_.size()),
+                    framebuffers_.data());
+
+  // Allocate initial images.
+  image_views_.resize(image_count);
+  Resize(size_);
+}
+
+ES3PlatformSwapChain::~ES3PlatformSwapChain() {
+  ES3PlatformContext::ThreadLock context_lock(
+      ES3PlatformContext::AcquireThreadContext(platform_context_));
+  glDeleteFramebuffers(static_cast<GLsizei>(framebuffers_.size()),
+                       framebuffers_.data());
+}
+
+SwapChain::ResizeResult ES3PlatformSwapChain::Resize(Size2D new_size) {
+  WTF_SCOPE0("ES3PlatformSwapChain#Resize");
+  ES3PlatformContext::ThreadLock context_lock(platform_context_);
+
+  // Recreate the underlying surface.
+  auto recreate_surface_result = platform_context_->RecreateSurface(new_size);
+  switch (recreate_surface_result) {
+    case ES3PlatformContext::RecreateSurfaceResult::kSuccess:
+      // OK.
+      break;
+    case ES3PlatformContext::RecreateSurfaceResult::kInvalidTarget:
+      LOG(ERROR) << "Failed to recreate swap chain surface; invalid target";
+      return ResizeResult::kDeviceLost;
+    case ES3PlatformContext::RecreateSurfaceResult::kOutOfMemory:
+      LOG(ERROR) << "Failed to recreate swap chain surface; out of memory";
+      return ResizeResult::kOutOfMemory;
+    case ES3PlatformContext::RecreateSurfaceResult::kDeviceLost:
+      LOG(ERROR) << "Failed to recreate swap chain surface; device lost";
+      return ResizeResult::kDeviceLost;
+  }
+
+  // Query the new size, as it may be different than requested.
+  size_ = platform_context_->QuerySize();
+  image_create_params_.size = Size3D(size_);
+
+  // Resize all images by recreating them.
+  for (int i = 0; i < image_views_.size(); ++i) {
+    image_views_[i].reset();
+  }
+  memory_pool_->Reclaim();
+  for (int i = 0; i < image_views_.size(); ++i) {
+    // Allocate image.
+    ref_ptr<Image> image;
+    auto result = memory_pool_->AllocateImage(image_create_params_, &image);
+    DCHECK(result == MemoryPool::AllocationResult::kSuccess);
+    if (!image) {
+      return ResizeResult::kOutOfMemory;
+    }
+
+    // Get a view for the target format.
+    image_views_[i] =
+        image->CreateView(Image::Type::k2D, image_create_params_.format);
+  }
+
+  return ResizeResult::kSuccess;
+}
+
+SwapChain::AcquireResult ES3PlatformSwapChain::AcquireNextImage(
+    std::chrono::milliseconds timeout_millis,
+    ref_ptr<QueueFence> signal_queue_fence,
+    ref_ptr<ImageView>* out_image_view) {
+  WTF_SCOPE0("ES3PlatformSwapChain#AcquireNextImage");
+  ES3PlatformContext::ThreadLock context_lock(platform_context_);
+
+  // NOTE: we always signal the fence right away, as we don't support overlap.
+  signal_queue_fence.As<ES3QueueFence>()->event()->Set();
+
+  int image_index = next_image_index_;
+  next_image_index_ = (next_image_index_ + 1) % image_views_.size();
```

Comment:
This will fail if ``image_views_.size()`` is ``0``. Can you add a sanity check here and/or in the constructor (if ``image_count`` *must* be greater than ``0`` - there's a call to ``glGenFramebuffers`` with image_count in the constructor so yes?).